components/director/internal/tenantfetcher/service.go

```diff
@@ -88,84 +154,103 @@ func (s Service) SyncTenants() error {
 	}
 
 	tenantsToCreate, err := s.getTenantsToCreate(lastConsumedTenantTimestamp)
-
 	if err != nil {
 		return err
 	}
-	tenantsToCreate = s.dedupeTenants(tenantsToCreate)
 
 	tenantsToDelete, err := s.getTenantsToDelete(lastConsumedTenantTimestamp)
 	if err != nil {
 		return err
 	}
 
-	totalNewEvents := len(tenantsToCreate) + len(tenantsToDelete)
+	runtimesToMove, err := s.getRuntimesToMoveByLabel(lastConsumedTenantTimestamp)
+	if err != nil {
+		return err
+	}
+
+	totalNewEvents := len(tenantsToCreate) + len(tenantsToDelete) + len(runtimesToMove)
 	log.C(ctx).Printf("Amount of new events: %d", totalNewEvents)
 	if totalNewEvents == 0 {
 		return nil
 	}
 
-	deleteTenantsMap := make(map[string]model.BusinessTenantMappingInput)
-	for _, ct := range tenantsToDelete {
-		deleteTenantsMap[ct.ExternalTenant] = ct
-	}
-
-	for i := len(tenantsToCreate) - 1; i >= 0; i-- {
-		if _, found := deleteTenantsMap[tenantsToCreate[i].ExternalTenant]; found {
-			tenantsToCreate = append(tenantsToCreate[:i], tenantsToCreate[i+1:]...)
-		}
-	}
+	tenantsToCreate = s.dedupeTenants(tenantsToCreate)
+	tenantsToCreate = s.excludeTenants(tenantsToCreate, tenantsToDelete)
 
 	tx, err := s.transact.Begin()
 	if err != nil {
 		return err
 	}
 	defer s.transact.RollbackUnlessCommitted(ctx, tx)
-
 	ctx = persistence.SaveToContext(ctx, tx)
 
-	currentTenants, err := s.tenantStorageService.List(ctx)
-	if err != nil {
-		return errors.Wrap(err, "while listing tenants")
-	}
+	existingTenantsSup := s.newCurrentTenantsSupplier()
```

Comment:
Is there a reason for this additional abstraction `currentTenantsSupplier`? As far as I see it's only used so that `createTenants` and `deleteTenants` can call it's `Get(ctx)` method. Why not simply make the `Get(ctx)` method a standalone function which returns existing tenants here and then provide them to the `createTenants` and `deleteTenants`?

Reply:
There were a meaning, but after all the refactoring maybe we don't need it anymore :D